### PR TITLE
docs: DEP-0007 for multi-head models

### PR DESCRIPTION
**Pull request by DaniilSergeev17**

### CODE REVIEW

**Review by yaz0p — DISMISSED** [review]

File: deps/0007-dep-multi-head-model-protocol.md [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+    def infer_stage_inputs_from_pipeline_inputs(self, inputs, n_microbatches):
+        ...
+
+    def infer_stage_outputs_from_pipeline_inputs(self, inputs, n_microbatches):
+        ...
+```
+
+`ClassificationModel` and `EmbeddingModel` follow the same shape, differing only in the four cells of the table above (head class + attr, the `pooling_mask` kwarg, the output key, and the inferred shape). The `infer_stage_inputs_from_pipeline_inputs` delegation and the backbone storage/forward/reset wiring are identical across all three.
+
+### Concrete models become thin subclasses
```
Comment (mrapplexz) [inline deps/0007-dep-multi-head-model-protocol.md]: These `*Model*For*Task*` classes will still be duplicated, even though they'll contain less code. This wouldn't solve a combinatorial explosion problem. How about deprecating these classes and making the users instantiate the models by composition manually like `model = CausalLMModel(Qwen3MoEModel(...))` ? Even better, since we are still at version 0.*, we can just remove these classes and break backward compatibility (we won't be able to do it after releasing 1.0).

**Review by mrapplexz — CHANGES_REQUESTED** [review]

**Comment by mrapplexz** [comment]
Another thing to note - this DEP is named `Multi-Head Model Protocol`, but currently it provides only models equipped with specific heads, not mentioning a "multi-head" model that, for instance, could provide multiple heads of one type or multiple heads of different types.

File: deps/0007-dep-multi-head-model-protocol.md (hunk 1 above)
Comment (DaniilSergeev17) [reply deps/0007-dep-multi-head-model-protocol.md]: It's great that we can break bc now, then I will delete these classes and substitute them with config-chosen heads

File: deps/0007-dep-multi-head-model-protocol.md [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+
+## Abstract
+
+Every architecture in `d9d` ships three near-identical task wrappers — `*ForCausalLM`, `*ForClassification`,
+`*ForEmbedding` — that share a backbone and differ only in their head. They are copy-pasted across families, so the
+duplication scales as `3 × N`. Each wrapper also hard-binds *exactly one* head, so a model with two heads (multi-task,
+or a reward head beside an LM head) cannot be expressed at all.
+
+This proposal replaces the wrappers with one composition primitive: a `DecoderBackbone` protocol, a `TaskHead` contract,
+and a single generic `DecoderModel` that composes one backbone with a *mapping* of named heads built from a closed,
```
Comment (mrapplexz) [inline deps/0007-dep-multi-head-model-protocol.md]: nit: on naming - in HF we have `SomeModel` for a regular model with no head attached; maybe we should follow a similar convention? like `DecoderModel` is a backbone, and `DecoderTask`/`DecoderWithHead`/... is a composed backbone with head attached to it

Comment (DaniilSergeev17) [reply deps/0007-dep-multi-head-model-protocol.md]: yes, sounds good. WDYT about DecoderWithHeads instead of DecoderModel? 

Also, I don't think we should name DecoderTask, because in d9d it may confuse with BaseTask (confusion: model & train task)

File: deps/0007-dep-multi-head-model-protocol.md [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+  `pooling_mask`) arrive via `**head_inputs`, and each head reads what it needs. An open set of heads cannot be named
+  kwargs — this mapping is the irreducible cost of genuine multi-head, and it documents each input on its consumer.
+
+## Usage
+
+A provider builds the backbone, then composes heads from config:
+
+```python
+backbone = Qwen3MoEModel(cfg.model, stage, hidden_states_snapshot_mode=..., enable_checkpointing=...)
+model = DecoderModel.from_configs(backbone, head_configs={"lm": CausalLMHeadConfig()}, stage=stage)
```
Comment (mrapplexz) [inline deps/0007-dep-multi-head-model-protocol.md]: I guess that DecoderModel should not be responsible for constructing heads - it should accept both the backbone and model as prebuilt modules. The backbone should be built the way you did - `backbone = Qwen3MoEModel(cfg.model, stage, hidden_states_snapshot_mode=..., enable_checkpointing=...)`. And the heads should be constructed using some another factory (if we talk about d9d-provided heads) or just passed directly if it's some head that is constructed by end-user

File: deps/0007-dep-multi-head-model-protocol.md [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+```
+
+A second head is a one-line change; the task then reads both keys:
+
+```python
+head_configs={"lm": CausalLMHeadConfig(), "cls": ClassificationHeadConfig(num_labels=3)}
+# loss = lm_loss(out["logps"]) + alpha * cls_loss(out["scores"])
```
Comment (mrapplexz) [inline deps/0007-dep-multi-head-model-protocol.md]: If I understood this correctly, the outputs are named without mentioning head names. Is this an intended behavior? What if the user wants to have multiple classification heads?

File: deps/0007-dep-multi-head-model-protocol.md (hunk 2 above)
Comment (mrapplexz) [reply deps/0007-dep-multi-head-model-protocol.md]: It's good, but a bit verbose. If we can come up with something simpler it would be great, but if not, it's okay

File: deps/0007-dep-multi-head-model-protocol.md (hunk 1 above)
Comment (mrapplexz) [reply deps/0007-dep-multi-head-model-protocol.md]: good!

**Comment by mrapplexz** [comment]
I have some general questions:

1. What will happen with multiple `parallelize_*` functions? They also suffer from all the repetition things this DEP is trying to solve 
2. The same question, but for HuggingFace mappers

File: deps/0007-dep-multi-head-model-protocol.md (hunk 2 above)
Comment (DaniilSergeev17) [reply deps/0007-dep-multi-head-model-protocol.md]: Alright, I decided to keep 'DecoderWithHeads', another variant 'HeadedDecoder' seems not to reflect it's functionality

File: deps/0007-dep-multi-head-model-protocol.md (hunk 4 above)
Comment (DaniilSergeev17) [reply deps/0007-dep-multi-head-model-protocol.md]: no, it isn't intended behavior, will fix it to out['lm/logps'] and out['cls/scores']

File: deps/0007-dep-multi-head-model-protocol.md [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+    def forward(self, hidden_states: torch.Tensor, inputs: Mapping[str, torch.Tensor | None]) -> dict[str, torch.Tensor]: ...
+
+    @abc.abstractmethod
+    def infer_output_shapes(self, pipeline_inputs: dict[str, torch.Tensor], n_microbatches: int) -> dict[str, torch.Tensor]: ...
+
+    @abc.abstractmethod
+    def parallelize(self, dist_context: DistributedContext) -> None: ...
+
+    @abc.abstractmethod
+    def huggingface_mapper(self, prefix: str) -> ModelStateMapper: ...
```
Comment (mrapplexz) [inline deps/0007-dep-multi-head-model-protocol.md]: Having the parallelization and HF methods inside a single TaskHead couples three different concerns together: regular head execution (fine to be inside TaskHead module), horizontal parallelism and compat mapping.  I believe these should be separated

File: deps/0007-dep-multi-head-model-protocol.md [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+    def __init__(self, backbone: TBackbone, heads: Mapping[str, TaskHead], stage: PipelineStageInfo): ...
+```
+
+The backbone is stored at `self.model` (FQN `model.*`); on the last stage the heads live in an `nn.ModuleDict` (FQN
+`heads.<name>.*`). `forward` runs the backbone, then each head over the hidden states, merging each head's outputs under
+its mapping key; `infer_stage_outputs` unions the heads' declared shapes; `infer_stage_inputs` delegates to the backbone.
+
+### Multi-head semantics
+
+Supporting more than one head follows from the flat `dict[str, torch.Tensor]` the pipeline already passes to the task:
```
Comment (mrapplexz) [inline deps/0007-dep-multi-head-model-protocol.md]: I'd like to add a comment here that sounds like "currently we should use a flat dictionary with keys simulating hierarchy mainly because it is expected by current PP engine implementation, but further we will consider migrating to a proper nested data structures for model outputs"

Comment (DaniilSergeev17) [reply deps/0007-dep-multi-head-model-protocol.md]: good concern

File: deps/0007-dep-multi-head-model-protocol.md (hunk 5 above)
Comment (DaniilSergeev17) [reply deps/0007-dep-multi-head-model-protocol.md]: Sounds reasonable, will do it in next commit

File: d9d/module/parallelism/model/head.py [hunk 7]
```diff
@@ -0,0 +1,17 @@
+from d9d.core.dist_context import DENSE_DOMAIN, DistributedContext
+from d9d.module.block.head import TaskHead
+from d9d.module.parallelism.api import parallelize_hsdp
+
+
+def parallelize_task_head(head: TaskHead, dist_context: DistributedContext) -> None:
```
Comment (mrapplexz) [inline d9d/module/parallelism/model/head.py]: I believe we shouldn't have a single "universal" method for all the task heads actually, because there will be different task head parallelization strategies in future. For instance, we may add vocab parallel for causal LM head in future

File: d9d/module/model/qwen3_dense/huggingface.py
Comment (mrapplexz) [inline d9d/module/model/qwen3_dense/huggingface.py:1]: Hm, regarding all the huggingface conversion stuff - currently HF only supports models with a single head attached. Moreover, all of these HF models have their own implementations and therefore parameter names for head parameters. That said, I believe we should leave previous "multiple hard-coded converter" approach as-is, but provide some way for the end user to control the name of a d9d model's head that will accept parameter from a huggingface model (and include some default name). It will be useful for one specific use-case, where the user wants to initialize a multi-head model from a single-head one - like the user may want to load causal LM head state from HF checkpoint, and initialize some other head randomly.

File: d9d/module/model/qwen3_moe/model.py [hunk 8]
```diff
@@ -100,9 +85,13 @@ def __init__(
 
         self._stage = stage
         self._hidden_states_snapshot_mode = hidden_states_snapshot_mode
-        self._hidden_size = params.layer.hidden_size
         self._enable_checkpointing = enable_checkpointing
 
+        # public backbone-shared dimensions the composed task heads derive from
+        self.hidden_size = params.layer.hidden_size
+        self.split_vocab_size = params.split_vocab_size
+        self.split_vocab_order = params.split_vocab_order
```
Comment (mrapplexz) [inline d9d/module/model/qwen3_moe/model.py]: If these are intended to be accessible from the outside, I'd prefer to make them read-only properties for more transparency

Comment (mrapplexz) [reply d9d/module/model/qwen3_moe/model.py]: and to state that they are immutable after construction

File: d9d/module/model/io/sequence.py [hunk 9]
```diff
@@ -45,30 +46,49 @@ class SequenceShared:
...
         sequence: The backbone shared input.
-        pooling_mask: Binary mask indicating which token(s) to pool (used on the last stage). You can
-            use ``d9d.dataset.token_pooling_mask_from_attention_mask`` to build it from an attention mask.
+        heads: Each head's own shared input, keyed by the names the heads were composed under
+            (read on the last stage).
     """
 
     sequence: SequenceShared
-    pooling_mask: torch.Tensor | None = None
+    heads: Mapping[str, Any]
```
Comment (mrapplexz) [inline d9d/module/model/io/sequence.py]: Making the head parameters `Any` could break the static type checking even for some default use cases like  single-head models. Could we define the contracts here in a stricter way?

File: d9d/module/block/head/huggingface.py
Comment (mrapplexz) [inline d9d/module/block/head/huggingface.py:1]: I left some comments about universal mappers for Qwens - they are related to this code too

File: d9d/module/block/head/factory.py [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+from typing import TYPE_CHECKING, Annotated, Literal
+
+from pydantic import BaseModel, Field
+
+from d9d.module.block.head.base import TaskHead
+from d9d.module.block.head.classification import ClassificationHead
+from d9d.module.block.head.embedding import EmbeddingHead
+from d9d.module.block.head.language_modelling import SplitLanguageModellingHead
+
+if TYPE_CHECKING:
```
Comment (mrapplexz) [inline d9d/module/block/head/factory.py]: this one is not required if you use `from __future__ import annotations`

File: d9d/module/block/head/factory.py [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+
+
+AnyHeadConfig = Annotated[
+    CausalLMHeadConfig | ClassificationHeadConfig | EmbeddingHeadConfig,
+    Field(discriminator="kind"),
+]
+"""Closed, discriminated union of the built-in head configurations."""
+
+
+def build_head(config: AnyHeadConfig, *, backbone: DecoderBackbone, stage: PipelineStageInfo) -> TaskHead:
```
Comment (mrapplexz) [inline d9d/module/block/head/factory.py]: `stage` is accepted but is unused

Comment (mrapplexz) [reply d9d/module/block/head/factory.py]: also, I guess this method's name should be somehow related to `DecoderBackbone`'s name - since it only works for DecoderBackbone, not for any other backbone (or it may?)

File: d9d/module/model/decoder.py [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+        Returns:
+            A ``SequenceTransfer`` of ``TensorSpec``.
+        """
+        ...
+
+
+TBackbone = TypeVar("TBackbone", bound=DecoderBackbone)
+
+
+class DecoderWithHeads(
```
Comment (mrapplexz) [inline d9d/module/model/decoder.py:45]: Shouldn't we add default implementations for single-head models like `DecoderForCausalLM`, `DecoderForClassification` and so on? It will simplify the construction for the most common use cases.

Comment (mrapplexz) [reply d9d/module/model/decoder.py:45]: also it may help with the `dict[str, Any]` for the head states

**Review by mrapplexz — CHANGES_REQUESTED** [review]

File: d9d/module/block/head/factory.py (hunk 11 above)
Comment (DaniilSergeev17) [reply d9d/module/block/head/factory.py]: I suppose build_decoder_head will be ok

File: d9d/module/model/qwen3_moe/model.py (hunk 8 above)
Comment (DaniilSergeev17) [reply d9d/module/model/qwen3_moe/model.py]: good idea

File: d9d/module/parallelism/model/head.py (hunk 7 above)
Comment (DaniilSergeev17) [reply d9d/module/parallelism/model/head.py]: Yes, we tried to be universal, but it seems a real concern here, will fix it

File: d9d/module/model/decoder.py [hunk 13]
```diff
@@ -0,0 +1,247 @@
...
+            pipeline_input: A representative ``SequenceInput`` microbatch; only shapes are read.
+            boundary: Which inter-stage edge to describe.
+
+        Returns:
+            A ``SequenceTransfer`` of ``TensorSpec``.
+        """
+        return self.model.stage_transfer_spec(pipeline_input, boundary)
+
+
+class DecoderForCausalLM(DecoderWithHeads[TBackbone, SequenceCausalLMHeadShared, SequenceCausalLMOutput]):
```
Comment (mrapplexz) [inline d9d/module/model/decoder.py]: should these classes really subclass DecoderWithHeads?

Comment (mrapplexz) [reply d9d/module/model/decoder.py]: it may help to remove additional nesting for outputs and remove the configurable head name

**Review by mrapplexz — APPROVED** [review]
